internal/netclassify/classify_test.go

```diff
@@ -0,0 +1,197 @@
+			want: "pkg:alpine/some_package@1.0.0-r0",
+		},
+		{
+			name:    "invalid alpine URL format",
+			url:     "https://dl-cdn.alpinelinux.org/alpine/v3.19/main/x86_64/invalid_format.apk",
+			want:    "",
+			wantErr: ErrUnclassified,
+		},
+
+		// PyPI URL tests
+		{
+			name: "valid PyPI wheel URL",
+			url:  "https://files.pythonhosted.org/packages/84/c2/80633736cd183ee4a62107413def345f7e6e3c01563dbca1417363cf957e/build-1.2.2.post1-py3-none-any.whl",
+			want: "pkg:pypi/build@1.2.2.post1",
+		},
+		{
+			name:    "PyPI metadata URL",
+			url:     "https://files.pythonhosted.org/packages/84/c2/80633736cd183ee4a62107413def345f7e6e3c01563dbca1417363cf957e/build-1.2.2.post1-py3-none-any.whl.metadata",
+			want:    "",
+			wantErr: ErrSkipped,
+		},
+		{
+			name:    "PyPI API URL",
+			url:     "https://pypi.org/simple/build/",
+			want:    "",
+			wantErr: ErrSkipped,
+		},
+		{
+			name: "PyPI source distribution URL",
+			url:  "https://files.pythonhosted.org/packages/84/c2/80633736cd183ee4a62107413def345f7e6e3c01563dbca1417363cf957e/sample-1.0.0.tar.gz",
+			want: "pkg:pypi/sample@1.0.0",
+		},
+		{
+			name:    "invalid PyPI URL",
+			url:     "https://files.pythonhosted.org/packages/invalid/format",
+			want:    "",
+			wantErr: ErrUnclassified,
+		},
+
+		// gcs URL tests
+		{
+			name: "valid GCS URL",
+			url:  "https://foo.storage.googleapis.com/bar/baz",
+			want: "pkg:generic/gcs/foo/bar/baz",
+		},
+		{
+			name: "valid xml GCS URL",
+			url:  "https://storage.googleapis.com/download/storage/v1/b/foo/o/bar/baz",
+			want: "pkg:generic/gcs/foo/bar/baz",
+		},
+
+		// Other cases
+		{
+			name:    "unknown URL",
+			url:     "https://example.com/invalid",
+			want:    "",
+			wantErr: ErrUnclassified,
+		},
+		{
+			name:    "empty URL",
+			url:     "",
+			want:    "",
+			wantErr: ErrUnclassified,
+		},
+	}
+
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			got, err := ClassifyURL(tt.url)
+			if tt.wantErr != nil {
+				if err == nil {
+					t.Errorf("ClassifyURL() error = nil, wantErr %v", tt.wantErr)
+				} else if !errors.Is(err, tt.wantErr) {
+					t.Errorf("ClassifyURL() error = %v, wantErr %v", err, tt.wantErr)
+				}
+			} else if err != nil {
+				t.Errorf("ClassifyURL() unexpected error = %v", err)
+			} else if got != tt.want {
+				t.Errorf("ClassifyURL() = %v, want %v", got, tt.want)
+			}
+		})
+	}
+}
+
+func TestClassifyPyPIFile(t *testing.T) {
+	tests := []struct {
+		name    string
+		fname   string
+		want    string
+		wantErr error
+	}{
+		{
+			name:  "valid wheel file",
+			fname: "package-1.0.0-py3-none-any.whl",
+			want:  "pkg:pypi/package@1.0.0",
+		},
+		{
+			name:    "metadata file",
+			fname:   "package-1.0.0-py3-none-any.whl.metadata",
+			wantErr: ErrSkipped,
+		},
+		{
+			name:    "egg file",
+			fname:   "package-1.0.0.egg",
+			wantErr: ErrUnclassified,
+		},
+		{
+			name:  "source distribution tar.gz",
+			fname: "package-1.0.0.tar.gz",
+			want:  "pkg:pypi/package@1.0.0",
```

Comment:
optional: Differentiate source vs non-source dependencies

Reply:
i think it's valuable in the future but I'm not sure whether it's critical right now. I've added a TODO and if/when this comes back from downstream consumers, we should revisit.